Η παρακάτω αναφορά έγινε με την επιμέλεια των φοιτητών Κριθαρούλα Αναστασίας με αριθμό μητρώου 03117073 και Κριθαρούλα Διονύση με αριθμό μητρώου 03117875.

# Τεχνητή Νοημοσύνη: Εργαστηριακή Άσκηση 2
---

Ο στόχος της εργασίας είναι η κατασκευή ενός συστήματος προτάσεων (Recommendation System) για ταινίες. Οι προτάσεις αυτές θα πηγάζουν τόσο από τα χαρακτηριστικά της ταινίας όσο και από ορισμένες αξιολογήσεις του κάθε χρήστη.

Στα δεδομένα της άσκησης περιλαμβάνονται ένα αρχείο με το όνομα movies_metadata.csv το όποιο περιέχει τα χαρακτηριστικά κάθε ταινίας όπως θέμα, σκηνοθέτης ηθοποιοί, λέξεις κλειδιά κ.α. από το imdb καθώς και τα αρχεία ratings.csv τα όποια περιέχουν πραγματικές αξιολογήσεις χρηστών, χωρισμένες σε train και σε test.


# Εκφώνηση
Στην παρούσα εργασία σας ζητείται να μελετήσετε και να υλοποιήσετε τα παρακάτω:<br>

### Μέρος 1
Στο μέρος 1 και 2 θα ασχοληθείτε και θα εργαστείτε μόνο με το αρχείο movies_metadata.csv.
#### Ερώτημα 1α

Αρχικά θα πρέπει, αφού μελετήσετε τη δομή και τα χαρακτηριστικά του αρχείου αυτού, να κατασκευάσετε μια βάση γνώσης για την Prolog η όποια ουσιαστικά θα αποτελεί τον κόσμο με τον όποιο θα εργαστείτε στην συνέχεια. Τα κατηγορήματα που θα δημιουργηθούν θα σας βοηθήσουν και στην κατασκευή του recommender και θα είναι της μορφής:

```
director(Movie, Director).
genre(Movie, Genre).
```
#### Ερώτημα 1β

Αφού δημιουργήσετε τον κόσμο του προβλήματος, στη συνέχεια καλείστε να δημιουργήσετε, σε Prolog, απλούς κανόνες οι οποίοι θα βρίσκουν όλες τις ταινίες με:
1.	Κοινό θέμα (κάποιες λέξεις σχετικά με το genre κοινές)
2.	Αρκετά κοινό θέμα (κάποιες λιγότερες λέξεις σχετικά με το genre κοινές π.χ. 3)
3.	Σχετικά κοινό θέμα (λίγες λέξεις σχετικές με το genre κοινές π.χ. 1)
4.	Κοινός σκηνοθέτης
5.	Ακριβώς ίδια πλοκή (κάποιες λέξεις κλειδιά της πλοκής κοινές)
6.	Σχετικά ίδια πλοκή (κάποιες λιγότερες λέξεις κλειδιά κοινές)
7.	Ίδιους τους βασικούς ηθοποιούς (και τους 3)
8.	Αρκετά ίδιους βασικούς ηθοποιούς (ορισμένους βασικούς ηθοποιούς κοινούς π.χ. 2)
9.	Σχετικά ίδιους ηθοποιούς (π.χ. 1 από τους 3)
10.	Ίδια γλώσσα
11.	Είναι έγχρωμες ή ασπρόμαυρες

Αξίζει να σημειωθεί ότι στα παραπάνω μπορείτε να προσθέσετε περισσότερα ερωτήματα ή να αλλάξετε την κλιμακωσιμότητα των queries (πέρα από το ίδιο, αρκετά ίδιο, σχετικά ίδιο) που θα κατασκευάσετε(π.χ. μια κλίμακα από 1 στα 5 όπου αυτό είναι δυνατόν), μιας και αυτά στην συνέχεια θα χρησιμοποιηθούν για την λειτουργία του recommender. Έτσι μπορείτε να προσθέσετε queries με τα όποια μπορεί να παράγονται καλύτερες συστάσεις  (το αρχείο movie_metadata.csv περιέχει πολλές πληροφορίες ακόμα για κάθε ταινία όπως έτος κυκλοφορίας, βαθμολογία στο imdb, facebοok_likes κ.α.).

Περισσότερες πληροφορίες για το dataset μπορείτε να διαβάσετε σε αυτό το [Link](https://www.kaggle.com/georgefila/movies-metadata):
### Μέρος 2: Recommendation System
Στο σημείο αυτό καλείστε με βάση αυτά που κάνατε στο μέρος 1 να κατασκευάσετε queries τα όποια θα σας επιστρέφουν παρόμοιες (σε χαρακτηριστικά) ταινίες. Τα ερωτήματα αυτά θα είναι κλιμακούμενα, δηλαδή θα υπάρχουν ερωτήματα που επιστρέφουν αρκετά κοινές ταινίες αλλά και που επιστρέφουν λιγότερο και λιγότερο κοινές (σε μια κλίμακα π.χ. από 1 σε 5). Για παράδειγμα:

```
find_simmilar_movies_5("Pirates Of The Caribbean", M).
M = "Pirates Of The Caribbean: On Stranger Tides"
M = "The Chronicles Of Narnia"
M = "Prince Of Persia: The Sands Of Time"
...
```
Για παράδειγμα, το παραπάνω ερώτημα θα επιστρέφει αρκετά κοινές σε περιεχόμενο ταινίες με την ταινια "Pirates Of The Caribbean". Θα υπάρχουν και αντίστοιχα ερωτήματα που θα βρίσκουν λιγότερο όμοιες ταινίες. Ο δείκτης ομοιότητας των ταινιών είναι αυθαίρετος και μπορείτε να τον ορίσετε εσείς όπως θέλετε, αρκεί να υπάρχει κάποια λογική σύνδεση με τα δεδομένα που περιέχονται στο αρχείο movies_metadata.csv. 

Συνεπώς η συνάρτηση που θα κάνει τις προτάσεις (recommendation) με είσοδο μια ταινία πρέπει να επιστρέφει-εκτυπώνει μια λίστα με τις προτεινόμενες ταινίες κατά φθίνουσα σειρά ομοιότητας. 


###3ο Μέρος: Recommendation System Με βάση τις προτιμήσεις - Αξιολογήσεις του χρήστη


Σε αυτό το σημείο θα εργαστείτε με τα αρχεία ratings τα όποια περιέχουν αξιολογήσεις (από 1 μέχρι 5) για τις παραπάνω ταινίες. Το προηγούμενο σύστημα συστάσεων προτείνει στον χρήστη ταινίες αποκλειστικά με βάση την ομοιότητά τους. Σε αυτό το σημείο θα γίνει μια αναβάθμιση του συστήματος έτσι ώστε να παράγονται καλύτερες  συστάσεις οι όποιες θα λαμβάνουν υπόψιν και τις προτιμήσεις του χρήστη, οι όποιες θα εξάγονται από τις αξιολογήσεις που έχει κάνει μέχρι στιγμής. 

Η εκπαίδευση του recommender θα γίνεται ως εξής:

Για κάθε ταινία θα υπάρχει ένα score το όποιο αρχικά θε είναι ίσο με 0 και θα διαμορφώνεται από τις αξιολογήσεις κάθε user. Έτσι για έναν χρήστη με βάση τις αξιολογήσεις που υπάρχουν στο αρχείο train_ratings θα πρέπει:

1.	Για κάθε ταινία που έχει βαθμολογήσει να βρίσκονται οι κοινές ταινίες ανά κλίμακα και στο μέχρι τώρα σκορ κάθε παρόμοιας ταινίας θα προστίθεται ένα βάρος το όποιο θα μπορούσε να είναι το ποσοστό ομοιότητας της ταινίας (δηλαδή ένα βάρος για κάθε κλίμακα) επί τον βαθμό που έχει βάλει ο χρήστης για την αρχική ταινία. Στη συνέχεια, ανάλογα με το σκορ που έχει σχηματιστεί για κάθε ταινία, θα επιλέγεται αν αυτή θα μπορούσε να είναι προτεινόμενη για τον χρήστη.

Η λογική πίσω από την παραπάνω διαδικασία είναι ότι παρόμοιες ταινίες θα έχουν ανάλογο βαθμό. Για παράδειγμα αν ένας χρήστης έχει αξιολογήσει αρκετές  ταινίες οι όποιες είναι sci-fiction με 5/5 τότε μια ταίνια sci-fiction την όποια δεν έχει δεί λογικά θα του αρέσει και θα έπρεπε να την προτείνουμε.

Μετά την εκπαίδευση του συστήματος σας καλείστε να δοκιμάσετε τον recommender που κατασκευάσατε στην πράξη. Για τον σκοπό αυτό θα φορτώσετε το αρχείο train_ratings.csv όπου περιέχονται οι αξιολογήσεις του ίδιου χρήστη για άλλες ταινίες. Το σύστημά σας πρέπει να προβλέπει αν μια ταινία θα πρέπει να προταθεί στον χρήστη. Μια ταίνια θα έπρεπε να έχει προταθεί στον  χρήστη, αν έχει βαθμό μεγαλύτερο του 3. Συνεπώς για την επίβλεψη του συστήματός σας θα πρέπει για κάθε μια από τις ταινίες του αρχείου test_ratings.csv να επιστρέφεται 1 ή 0 αν η ταινία θα έπαιρνε βαθμό μεγαλύτερο του 3 ή όχι, δηλαδή αν θα έπρεπε να την είχαμε προτείνει ή όχι.

Στην συνέχεια, σε συνδιασμό με τις πραγματικές απαντήσεις του χρήστη θα αξιολογήσετε το σύστημά σας χρησιμοποιόντας τις μετρικές: precision, recall, f1 οι οποίες είναι οι πλέον γνωστές μετρικές και ευρέως χρησιμοποιούμενες τεχνικές για την επίβλεψη-μέτρηση απόδοσης ανάλογων συστημάτων.

1. Precision: Δείχνει πόσο ακριβές είναι το σύστημα. Υπολογίζει πόσα από τα στιγμιότυπα τα όποια προβλέψαμε ότι ανήκουν σε μια κλάση όντως ανήκουν σε αυτή. Η μετρική αυτή μας δίνει μια εικόνα σχετικά με τον αριθμό των ταινίων που προβλέψαμε ως προτεινόμενες ενώ δεν θα έπρεπε.

2. Recall: Υπολογίζει πόσα από τα στιγμιότυπα που ανήκουν σε μια κλάση (π.χ. προτεινόμενες ταινίες) προβλέφθηκαν σωστά.

3. F1: Είναι ένας μέσος μεταξύ των παραπάνω δυο μετρικών, έτσι ώστε να διατηρείται μια ισορροπία μεταξύ τους. Υπολογίζεται από την παρακάτω σχέση:

$$F_1=2\frac{Precision\times{Recall}}{Precision+Recall}$$

Οι παραπάνω συναρτήσεις παρέχονται από την βιβλιοθήκη scikit-learn.

Τέλος για την καλύτερη επίβλεψη του συστήματος σας μπορείτε να εκπαιδεύσετε  τον recommender σας με ένα υποσύνολο ταινίων (π.χ. 10, 20, 50, ...) για να μελετήσετε κατά πόσο σας βοηθούν οι επιπλέον αξιολογήσεις κάθε φορά (δηλαδή κατά πόσο βελτιώνονται οι παραπάνω μετρικές στο test set). Έτσι π.χ. μπορείτε να εντοπίσετε περιπτώσεις όπως για παράδειγμα ότι με έναν recommender μπορεί να μην επιτυγχάνετε πολύ υψηλό σκορ όσο με άλλους, αλλά το βέλτιστο σκόρ σας επιτυγχάνεται πολύ γρηγόρα π.χ. με μόνο 10 ταινίες αντί 100.



---
## Κατασκευή Περιβάλλοντος Εργασίας
---
### Διάβασμα Αρχείων στο Colab (Μόνο για το Colab)
Αν η υλοποίηση γίνει στο google colab τότε μπορεί να χρησιμοποιηθεί το google drive ως file system. Για να γίνει Mount το google drive τρέχουμε τον παρακάτω κώδικα και κλικάρουμε στο link που θα μας εμφανιστεί. 

```
from google.colab import drive
drive.mount('/gdrive')
import os
os.listdir('/gdrive/My Drive')
```
Έπειτα στην σελίδα που άνοιξε επιλέγουμε το mail μας και στο επόμενο παράθυρο που θα μας ανοίξει πατάμε Να επιτρέπεται. Στην συνέχεια αντιγράφουμε τον κωδικό που θα μας βγάλει και τον κάνουμε paste στο Input που έχει ανοίξει στο colab. Έτσι πλέον αν έχουμε ανεβάσει ένα αρχείο στο google drive μπορούμε να το βρούμε στην θέση:

```
movies_filename = '/gdrive/My Drive/' + movies_metadata.csv
```
Μπορούμε πλέον κανονικά να δουλέψουμε φτιάχνοντας φακέλους ή αρχεία και γενικότερα κάνοντας οτιδήποτε θα κάναμε αν ήμασταν τοπικά. 

### Prolog μέσω Python

Το πακέτο που θα χρησιμοποιηθεί για την επικοινωνία Python και Prolog είναι το pyswip (https://pypi.org/project/pyswip/). Για να δουλέψει το Pyswip  χρειάζεται να υπάρχει το Swi-Prolog το όποιο αν δουλεύουμε τοπικά πρέπει να το εγκαταστήσουμε, ακολουθώντας αντίστοιχες οδηγίες στην σελίδα του εργαλείου. Για να γίνει του Swi-Prolog η εγκατάσταση στο Google Colab πρέπει να τρέξουμε τον παρακάτω κώδικα:

```
!sudo apt-get install software-properties-common
!sudo apt-add-repository ppa:swi-prolog/stable
!sudo apt-get update
!sudo apt-get install swi-prolog
```
Σε κάποιο σημείο της εκτέλεσης εμφανίζεται ένα μήνυμα ότι πρέπει να πατήσουμε enter σε ένα input για να συνεχίσει η διαδικασία. Έπειτα από αυτό η εκτέλεση θα συνεχίσει χωρίς κάποιο πρόβλημα.

Τέλος πρέπει να εγκαταστήσουμε το pyswip (**όπου και να δουλεύουμε**) όπως παρακάτω:

```
!pip isntall pyswip
```


# **Μέρος 1: Μελέτη των Metadata, Δημιουργία κόσμου και των βασικών queries.**


In [1]:
import pandas as pd
from pyswip import Prolog

In [2]:
# Η βιβλιοθήκη pandas είναι χρήσιμη για την εργασία με τέτοια δεδομένα
import pandas as pd 
# Διάβασμα του αρχείου 'movie_metadata.csv' 
data = pd.read_csv("movies_metadata.csv") 
#Στο csv υπαρχούν κελία με nan τιμές
#Στις θέσεις αυτές βάζουμε 'UNK' πράγμα που  κάνουμε με την παρακάτω συνάρτηση
data.fillna("UNK", inplace=True)
# Preview the first 5 lines of the loaded data 
data.head()

Unnamed: 0     budget                                    genres  \
0           0  237000000  Action|Adventure|Fantasy|Science Fiction   
1           1  300000000                  Adventure|Fantasy|Action   
2           2  245000000                    Action|Adventure|Crime   
3           3  250000000               Action|Crime|Drama|Thriller   
4           4  260000000          Action|Adventure|Science Fiction   

                                       homepage      id  \
0                   http://www.avatarmovie.com/   19995   
1  http://disney.go.com/disneypictures/pirates/     285   
2   http://www.sonypictures.com/movies/spectre/  206647   
3            http://www.thedarkknightrises.com/   49026   
4          http://movies.disney.com/john-carter   49529   

                                       plot_keywords  language  \
0  culture clash|future|space war|space colony|so...   English   
1  ocean|drug abuse|exotic island|east india trad...   English   
2  spy|based on novel|secret agent|sequel|mi6|bri...  Français   
3  dc comics|crime fighter|terrorist|secret ident...   English   
4  based on novel|mars|medallion|space travel|pri...   English   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  popularity  ...  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577  ...   
1  Captain Barbossa, long believed to be dead, ha...  139.082615  ...   
2  A cryptic message from Bond’s past sends him o...  107.376788  ...   
3  Following the death of District Attorney Harve...  112.312950  ...   
4  John Carter is a war-weary, former military ca...   43.926995  ...   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   
2                           A Plan No One Escapes   
3                                 The Legend Ends   
4            Lost in our world, found in another.   

                                movie_title vote_average  num_voted_users  \
0                                    Avatar          7.2            11800   
1  Pirates of the Caribbean: At World's End          6.9             4500   
2                                   Spectre          6.3             4466   
3                     The Dark Knight Rises          7.6             9106   
4                               John Carter          6.1             2124   

  title_year                   country      director_name     actor_1_name  \
0       2009  United States of America      James Cameron      Zoe Saldana   
1       2007  United States of America     Gore Verbinski    Orlando Bloom   
2       2015            United Kingdom         Sam Mendes  Christoph Waltz   
3       2012  United States of America  Christopher Nolan    Michael Caine   
4       2012  United States of America     Andrew Stanton     Lynn Collins   

       actor_2_name       actor_3_name  
0  Sigourney Weaver       Stephen Lang  
1   Keira Knightley  Stellan Skarsgård  
2       Léa Seydoux      Ralph Fiennes  
3       Gary Oldman      Anne Hathaway  
4   Samantha Morton       Willem Dafoe  

[5 rows x 27 columns]

In [3]:
def clean_text(text):
    text = text.replace(u'\xa0', u'')
    text = text.replace(u"'", u'')
    return text

### Ερώτημα 1α - Δημιουργία του κόσμου του προβλήματος

In [4]:
#create World
#Ορίζουμε τον κόσμο μας
prolog = Prolog()

#Για κάθε row του πίνακα φτιάχνουμε τα κατηγορήματα που θέλουμε να αποθηκέυσουμε
#αρχικά σε μια λίστα με το όνομα literals
literals = []
movie_score = {}
for row in data.itertuples(index=True, name='Pandas'):
    
    movie_title = clean_text(getattr(row, 'movie_title'))
    identifier = str(getattr(row, 'id'))
    
    literals.append("identifier('"+ identifier +"','"+  movie_title +"')")
    
    literals.append("director('"+ identifier +"','"+  clean_text(getattr(row, 'director_name')) +"')")
    
    for genre in getattr(row, 'genres').split("|"):
        literals.append("genre('"+ identifier +"','"+ genre +"')")
    
    literals.append("actor('"+ identifier +"','"+  clean_text(getattr(row, 'actor_1_name')) +"')")
    literals.append("actor('"+ identifier +"','"+  clean_text(getattr(row, 'actor_2_name')) +"')")
    literals.append("actor('"+ identifier +"','"+  clean_text(getattr(row, 'actor_3_name')) +"')")
    
    
    flag = 1
    for plot in getattr(row, 'plot_keywords').split("|"):
        if plot == 'black and white':
            literals.append("black_and_white('"+ identifier +"')")
            flag = 0
        else :
            literals.append("plot('"+ identifier +"','"+ clean_text(plot) +"')")
    
    if flag == 1:
        literals.append("colored('"+ identifier +"')")
        
    
    literals.append("language('"+ identifier +"','"+  getattr(row, 'language') +"')")
    
       
    title_year = getattr(row, 'title_year')
    if title_year != 'UNK':
        if int(title_year) < 2000 :
            literals.append("old_movie('"+ identifier +"')")
        elif int(title_year) < 2010 :
            literals.append("semi_modern_movie('"+ identifier +"')")
        else :
            literals.append("modern_movie('"+ identifier +"')")
   
    duration = getattr(row, 'duration')
    if duration != 'UNK':
        if int(duration) < 40 :
            literals.append("sort_movie('"+ identifier +"')")
        elif int(duration) < 220 :
            literals.append("normal_duration_movie('"+ identifier +"')")
        else :
            literals.append("long_movie('"+ identifier +"')")            
    
    
    budget = getattr(row, 'budget')
    if int(budget) > 38000000 :
        literals.append("high_budget('"+ identifier +"')")
    else :
        literals.append("low_budget('"+ identifier +"')") 
        
    production_companies = getattr(row, 'production_companies')
    production_companies = production_companies.split('}')
    for p in production_companies[:len(production_companies) - 1]:
        literals.append("production_company('"+ identifier +"','"+ p[-1] +"')") 
        
    literals.append("country('"+ identifier +"','"+  clean_text(getattr(row, 'country')) +"')")
    

#Η Prolog θέλει τα κατηγορήματά της με την σειρά 
literals.sort()
for literal in literals:
    prolog.assertz(literal)

#Επίσης μπορούμε να κάνουμε consult ένα έτοιμο αρχείο στον κόσμο όπως παρακάτω
#prolog.consult("db.pl")

Αφού μελετήσαμε τη δομή και τα χαρακτηριστικά του αρχείου movies_metadata.csv αποφασίσαμε να κατασκευάσουμε μια βάση γνώσης για την Prolog με τα ακόλουθα κατηγορήματα:

- identifier( <i>movie_id</i> , <i>movie_title</i> )
- genre( <i>movie_id</i> , <i>genre</i> )
- plot( <i>movie_id</i> , <i> plot_keyword</i> )
- colored( <i>movie_id</i> ) : εάν η συγκεκριμένη ταινία με <i>movie_id</i> είναι έγχρωμη
- black_and_white( <i>movie_id</i> ) : εάν η συγκεκριμένη ταινία με <i>movie_id</i> είναι ασπρόμαυρη
- actor( <i>movie_id</i> , <i> actor_name</i> )
- language( <i>movie_id</i> , <i> language</i> )
- old_movie( <i>movie_id</i> ) : εάν η συγκεκριμένη ταινία με <i>movie_id</i> κυκλοφόρησε πριν το 2000
- semi_modern_movie( <i>movie_id</i> ) : εάν η συγκεκριμένη ταινία με <i>movie_id</i> κυκλοφόρησε μετά το 2000 και πριν 2010
- modern_movie( <i>movie_id</i> ) : εάν η συγκεκριμένη ταινία με <i>movie_id</i> κυκλοφόρησε μετά το 2010
- sort_movie( <i>movie_id</i> ) : εάν η συγκεκριμένη ταινία με <i>movie_id</i> έχει διάρκεια μικρότερη από 40 λεπτά
- normal_duration_movie( <i>movie_id</i> ) : εάν η συγκεκριμένη ταινία με <i>movie_id</i> έχει διάρκεια μεγαλύτερη από 40 λεπτά και μικρότερη από 220 λεπτά
- long_movie( <i>movie_id</i> ) : εάν η συγκεκριμένη ταινία με <i>movie_id</i> έχει διάρκεια μεγαλύτερη από 220 λεπτά
- high_budget( <i>movie_id</i> ) : εάν η συγκεκριμένη ταινία με <i>movie_id</i> έχει budget μεγαλύτερo από 38000000
- low_budget( <i>movie_id</i> ) : εάν η συγκεκριμένη ταινία με <i>movie_id</i> έχει budget μικρότερο από 38000000
- production_company( <i>movie_id</i> , <i> production_company_id</i> )
- country( <i>movie_id</i> , <i>country</i> )

### Ερώτημα 1β - Δημιουργία κανόνων

Εκμεταλλευόμενοι τα κατηγορήματα που ορίσαμε παραπάνω ορίζουμε κανόνες οι οποίοι στη συνέχεια θα μας βοηθήσουν να κατηγοριοποιήσουμε τις ταινίες με βάση την ομοιότητά τους. Οι κανόνες αυτοί είναι:   

- <b> same_genres_3(X, Y) </b> : αληθεύει όταν δύο ταινίες με id X και Y αντίστοιχα έχουν κοινά τουλάχιστον τρία διαφορετικά χαρακτηριστικά (genres).
- <b> same_genres_2(X, Y) </b> : αληθεύει όταν δύο ταινίες με id X και Y αντίστοιχα έχουν κοινά τουλάχιστον δύο διαφορετικά χαρακτηριστικά (genres).
- <b> same_genres_(X, Y) </b> : αληθεύει όταν δύο ταινίες με id X και Y αντίστοιχα έχουν κοινό τουλάχιστον ένα χαρακτηριστικό (genre).

In [5]:
prolog.assertz('(same_genres_3(X, Y) :- genre(X, G1), genre(Y, G1), X \= Y, genre(X, G2), genre(Y, G2), G1 \= G2, genre(X, G3), genre(Y, G3), G1 \= G3, G2 \= G3)')
prolog.assertz('(same_genres_2(X, Y) :- genre(X, G1), genre(Y, G1), X \= Y, genre(X, G2), genre(Y, G2), G1 \= G2)')
prolog.assertz('(same_genres_1(X, Y) :- genre(X, G1), genre(Y, G1), X \= Y)')

- <b> same_director(X, Y) </b> : αληθεύει όταν δύο ταινίες με id X και Y αντίστοιχα έχουν κοινό σκηνοθέτη.

In [6]:
prolog.assertz('(same_director(X, Y) :- director(X, D), director(Y, D), X \= Y)')

- <b> same_plot_3(X, Y) </b> : αληθεύει όταν δύο ταινίες με id X και Y αντίστοιχα έχουν κοινές τουλάχιστον τρεις διαφορετικές αντιπροσωπευτικές λέξεις για την πλοκή τους.
- <b> same_plot_3(X, Y) </b> : αληθεύει όταν δύο ταινίες με id X και Y αντίστοιχα έχουν κοινές τουλάχιστον δύο διαφορετικές αντιπροσωπευτικές λέξεις για την πλοκή τους.
- <b> same_plot_3(X, Y) </b> : αληθεύει όταν δύο ταινίες με id X και Y αντίστοιχα έχουν κοινή τουλάχιστον μια αντιπροσωπευτική λέξη για την πλοκή τους.

In [7]:
prolog.assertz('(same_plot_3(X, Y) :- plot(X, P1), plot(Y, P1), X \= Y, plot(X, P2), plot(Y, P2), P1 \= P2, plot(X, P3), plot(Y, P3), P1 \= P3, P2 \= P3)')
prolog.assertz('(same_plot_2(X, Y) :- plot(X, P1), plot(Y, P1), X \= Y, plot(X, P2), plot(Y, P2), P1 \= P2)')
prolog.assertz('(same_plot_1(X, Y) :- plot(X, P1), plot(Y, P1), X \= Y)')

- <b> same_actors_3(X, Y) </b> : αληθεύει όταν δύο ταινίες με id X και Y αντίστοιχα έχουν κοινούς τουλάχιστον τρεις διαφορετικούς ηθοποιούς.
- <b> same_actors_2(X, Y) </b> : αληθεύει όταν δύο ταινίες με id X και Y αντίστοιχα έχουν κοινούς τουλάχιστον δύο διαφορετικούς ηθοποιούς.
- <b> same_actors_1(X, Y) </b> : αληθεύει όταν δύο ταινίες με id X και Y αντίστοιχα έχουν κοινό τουλάχιστον έναν διαφορετικό ηθοποιό.

In [8]:
prolog.assertz('(same_actors_3(X, Y) :- actor(X, A1), actor(Y, A1), X \= Y, actor(X, A2), actor(Y, A2), A1 \= A2, actor(X, A3), actor(Y, A3), A1 \= A3, A2 \= A3)')
prolog.assertz('(same_actors_2(X, Y) :- actor(X, A1), actor(Y, A1), X \= Y, actor(X, A2), actor(Y, A2), A1 \= A2)')
prolog.assertz('(same_actors_1(X, Y) :- actor(X, A1), actor(Y, A1), X \= Y)')

- <b> same_language(X, Y) </b> : αληθεύει όταν δύο ταινίες με id X και Y αντίστοιχα έχουν κοινή γλώσσα.

In [9]:
prolog.assertz('(same_language(X, Y) :- language(X, L1), language(Y, L1), X \= Y)')

- <b> same_color(X, Y) </b> : αληθεύει όταν δύο ταινίες με id X και Y είναι και οι δύο ασπρόμαυρες ή και οι δύο έγχρωμες.

In [10]:
prolog.assertz("(same_color(X, Y) :- black_and_white(X), black_and_white(Y), X \= Y)")
prolog.assertz("(same_color(X, Y) :- colored(X), colored(Y), X \= Y)")

- <b> same_year(X, Y) </b> : αληθεύει όταν δύο ταινίες με id X και Y έχουν κυκλοφορήσει την ίδια περίοδο, δηλαδή είτε πριν το 2000, είτε μεταξύ του 2000 και του 2010 είτε μετά το 2010.

In [11]:
prolog.assertz("(same_year(X, Y) :- old_movie(X), old_movie(Y), X \= Y)")
prolog.assertz("(same_year(X, Y) :- semi_modern_movie(X), semi_modern_movie(Y), X \= Y)")
prolog.assertz("(same_year(X, Y) :- modern_movie(X), modern_movie(Y), X \= Y)")

- <b> same_duration(X, Y) </b> : αληθεύει όταν δύο ταινίες με id X και Y έχουν περίπου την ίδια χρονική διάρκεια, δηλαδή είτε μικρότερη από 40 λεπτά, είτε μεταξύ 40 και 220 λεπτά είτε μεγαλύτερη από 220 λεπτά.

In [12]:
prolog.assertz("(same_duration(X, Y) :- sort_movie(X), sort_movie(Y), X \= Y)")
prolog.assertz("(same_duration(X, Y) :- normal_duration_movie(X), normal_duration_movie(Y), X \= Y)")
prolog.assertz("(same_duration(X, Y) :- long_movie(X), long_movie(Y), X \= Y)")

- <b> same_budget(X, Y) </b> : αληθεύει όταν δύο ταινίες με id X και Y έχουν περίπου το ίδιο budget, δηλαδή είτε μικρότερο από 38000000, είτε μεγαλύτερο από 38000000.

In [13]:
prolog.assertz("(same_budget(X, Y) :- high_budget(X), high_budget(Y), X \= Y)")
prolog.assertz("(same_budget(X, Y) :- low_budget(X), low_budget(Y), X \= Y)")

- <b> same_production_company_2(X, Y) </b> : αληθεύει όταν δύο ταινίες με id X και Y αντίστοιχα έχουν κοινές δύο τουλάχιστον διαφορετικές εταιρίες παραγωγής.
- <b> same_production_company_1(X, Y) </b> : αληθεύει όταν δύο ταινίες με id X και Y αντίστοιχα έχουν κοινή μια τουλάχιστον εταιρία παραγωγής.

In [14]:
prolog.assertz('(same_production_company_2(X, Y) :- production_company(X, C1), production_company(Y, C1), X \= Y, production_company(X, C2), production_company(Y, C2), C1 \= C2)')
prolog.assertz('(same_production_company_1(X, Y) :- production_company(X, C), production_company(Y, C), X \= Y)')

- <b> same_language(X, Y) </b> : αληθεύει όταν δύο ταινίες με id X και Y αντίστοιχα διαδραματίστηκαν στην ίδια χώρα.

In [15]:
prolog.assertz('(same_country(X, Y) :- country(X, C), country(Y, C), X \= Y)')

# **2ο Μέρος: Recommendation System με βάση μόνο τα χαρακτηριστικά των ταινιών.**

Στο σημείο αυτό με βάση τους κανόνες που κατασκευάστηκαν στο Μέρος 1 θα κατασκευαστούν κατηγορήματα για την ομοιότητα ταινιών. Παρακάτω δίνεται ένα μικρό παράδειγμα ενός κανόνα και πώς αυτός θα μπορούσε να γραφτεί μέσω του Pyswip. Επίσης όπως αναφέρεται και σε σχόλιο παραπάνω θα μπορεί να γραφτεί και μια βάση δεδομένων με τους κανόνες και να γίνει απευθείας consult.

Στο παρακάτω παράδειγμα το 5 και το 4 εκφράζουν την ομοιότητα των ταινιών π.χ. οι ταινίες που παράγονται μέσω του find_similar_5 είναι πιο όμοιες από αυτές που παράγονται μέσω του find_similar_4. 


## Ορισμός επιπέδων ομοιότητας μεταξύ των ταινιών

Παρακάτω ορίζουμε για μεγαλύτερη ακρίβεια στο recommendation σύστημα που θα δημιουργήσουμε ορίζουμε δέκα επίπεδα ομοιότητα μεταξύ των ταινιών, με το πρώτο επίπεδο ( <i> find_similar_1 </i> ) να ορίζει τις λιγότερο όμοιες ταινίες και το τελευταίο επίπεδο ( <i> find_similar_10 </i> ) να ορίζει τις ταινίες με τη μεγαλύτερη δυνατή ομοιότητα στο σύστημά μας.

### 1$^ο$ Επίπεδο ομοιότητας

Δύο ταινίες ορίζονται όμοιες στο 1ο επίπεδο εάν έχουν ένα κοινό χαρακτηριστικό.

In [16]:
prolog.assertz('''(find_sim_1(X, Y):-
                                        identifier(Xid, X),
                                        same_genres_1(Xid, Yid),
                                        identifier(Yid, Y))''')


### 2$^ο$ Επίπεδο ομοιότητας

Δύο ταινίες ορίζονται όμοιες στο 2ο επίπεδο εάν έχουν:

 - ένα κοινό χαρακτηριστικό, έχουν κυκλοφορήσει την ίδια χρονική περίοδο και είναι είτε έγχρωμες είτε ασπρόμαυρες, ή

 - μια κοινή αντιπροσωπευτική λέξη για την πλοκή τους,έχουν κυκλοφορήσει την ίδια χρονική περίοδο και είναι είτε έγχρωμες είτε ασπρόμαυρες.

In [17]:
prolog.assertz('''(find_sim_2(X, Y):-
                                        identifier(Xid, X),
                                        (same_genres_1(Xid, Yid) ; same_plot_1(Xid, Yid)),
                                        same_year(Xid, Yid),
                                        same_color(Xid, Yid),
                                        identifier(Yid, Y))''')


### 3$^ο$ Επίπεδο ομοιότητας

Δύο ταινίες ορίζονται όμοιες στο 3ο επίπεδο εάν έχουν ένα κοινό χαρακτηριστικό, μια κοινή αντιπροσωπευτική λέξη για την πλοκή τους, έναν κοινό ηθοποιό, έχουν κυκλοφορήσει την ίδια χρονική περίοδο και είναι είτε έγχρωμες είτε ασπρόμαυρες.

In [18]:
prolog.assertz('''(find_sim_3(X, Y):-
                                        identifier(Xid, X),
                                        same_genres_1(Xid, Yid),
                                        same_actors_1(Xid, Yid),
                                        same_year(Xid, Yid),
                                        same_color(Xid, Yid),
                                        identifier(Yid, Y))''')

### 4$^ο$ Επίπεδο ομοιότητας

Δύο ταινίες ορίζονται όμοιες στο 4ο επίπεδο εάν έχουν ένα κοινό χαρακτηριστικό, μια κοινή αντιπροσωπευτική λέξη για την πλοκή τους, έχουν κυκλοφορήσει την ίδια χρονική περίοδο, είναι είτε έγχρωμες είτε ασπρόμαυρες, έχουν περίπου την ίδια χρονική διάρκεια και περίπου το ίδιο budget.

In [19]:
prolog.assertz('''(find_sim_4(X, Y):-
                                        identifier(Xid, X),
                                        same_genres_1(Xid, Yid),
                                        same_plot_1(Xid, Yid),
                                        same_year(Xid, Yid),
                                        same_color(Xid, Yid),
                                        same_budget(Xid, Yid),
                                        same_duration(Xid, Yid),
                                        identifier(Yid, Y))''')

### 5$^ο$ Επίπεδο ομοιότητας

Δύο ταινίες ορίζονται όμοιες στο 5ο επίπεδο εάν έχουν:

 - δύο κοινά χαρακτηριστικά, έχουν κυκλοφορήσει την ίδια χρονική περίοδο, είναι είτε έγχρωμες είτε ασπρόμαυρες και έχουν περίπου την ίδια χρονική διάρκεια, ή

 - δύο κοινές αντιπροσωπευτικές λέξεις για την πλοκή τους, έχουν κυκλοφορήσει την ίδια χρονική περίοδο, είναι είτε έγχρωμες είτε ασπρόμαυρες έχουν περίπου την ίδια χρονική διάρκεια.

In [20]:
prolog.assertz('''(find_sim_5(X, Y):-
                                        identifier(Xid, X),
                                        (same_genres_2(Xid, Yid) ; same_plot_2(Xid, Yid)),
                                        same_year(Xid, Yid),
                                        same_duration(Xid, Yid),
                                        same_color(Xid, Yid),
                                        identifier(Yid, Y))''')

### 6$^ο$ Επίπεδο ομοιότητας

Δύο ταινίες ορίζονται όμοιες στο 6ο επίπεδο εάν έχουν:

- δύο κοινά χαρακτηριστικά, μια κοινή αντιπροσωπευτική λέξη για την πλοκή τους, έχουν κυκλοφορήσει την ίδια χρονική περίοδο, είναι είτε έγχρωμες είτε ασπρόμαυρες και έχουν περίπου την ίδια χρονική διάρκεια, ή

- δύο κοινές αντιπροσωπευτικές λέξεις για την πλοκή τους, ένα κοινό χρακτηριστικό, έχουν κυκλοφορήσει την ίδια χρονική περίοδο, είναι είτε έγχρωμες είτε ασπρόμαυρες και έχουν περίπου την ίδια χρονική διάρκεια.

In [21]:
prolog.assertz('''(find_sim_6(X, Y):-
                                        identifier(Xid, X),
                                        (same_genres_2(Xid, Yid), same_plot_1(Xid, Yid) ; same_genres_1(Xid, Yid), same_plot_2(Xid, Yid)),
                                        same_year(Xid, Yid),
                                        same_duration(Xid, Yid),
                                        same_color(Xid, Yid),
                                        identifier(Yid, Y))''')


### 7$^ο$ Επίπεδο ομοιότητας

Δύο ταινίες ορίζονται όμοιες στο 7ο επίπεδο εάν έχουν:

 - τρια κοινά χαρακτηριστικά, έχουν κυκλοφορήσει την ίδια χρονική περίοδο ,είναι είτε έγχρωμες είτε ασπρόμαυρες και έχουν περίπου την ίδια χρονική διάρκεια, ή

 - τρεις κοινές αντιπροσωπευτικές λέξεις για την πλοκή τους, έναν κοινό ηθοποιό, έχουν κυκλοφορήσει την ίδια χρονική περίοδο ,είναι είτε έγχρωμες είτε ασπρόμαυρες και έχουν περίπου την ίδια χρονική διάρκεια.

In [22]:
prolog.assertz('''(find_sim_7(X, Y):-
                                        identifier(Xid, X),
                                        (same_genres_3(Xid, Yid) ; same_plot_3(Xid, Yid), same_actors_1(Xid, Yid)),
                                        same_year(Xid, Yid),
                                        same_color(Xid, Yid),
                                        same_duration(Xid, Yid),
                                        identifier(Yid, Y))''')

### 8$^ο$ Επίπεδο ομοιότητας

Δύο ταινίες ορίζονται όμοιες στο 8ο επίπεδο εάν έχουν:

 - τρια κοινά χαρακτηριστικά, μια κοινή αντιπροσωπευτική λέξη για την πλοκή τους, κοινό σκηνοθέτη, έχουν κυκλοφορήσει την ίδια χρονική περίοδο ,είναι είτε έγχρωμες είτε ασπρόμαυρες, έχουν περίπου την ίδια χρονική διάρκεια και έχουν διαδραματιστεί στην ίδια χώρα, ή

 - τρεις κοινές αντιπροσωπευτικές λέξεις για την πλοκή τους, ένα κοινό χαρακτηριστικό,κοινό σκηνοθέτη, έχουν κυκλοφορήσει την ίδια χρονική περίοδο ,είναι είτε έγχρωμες είτε ασπρόμαυρες, έχουν περίπου την ίδια χρονική διάρκεια και έχουν διαδραματιστεί στην ίδια χώρα.

In [23]:
prolog.assertz('''(find_sim_8(X, Y):-
                                        identifier(Xid, X),
                                        (same_genres_3(Xid, Yid), same_plot_1(Xid, Yid) ; same_genres_1(Xid, Yid), same_plot_3(Xid, Yid)),
                                        same_director(Xid, Yid),
                                        same_year(Xid, Yid),
                                        same_color(Xid, Yid),
                                        same_duration(Xid, Yid),
                                        same_country(Xid, Yid),
                                        identifier(Yid, Y))''')

### 9$^ο$ Επίπεδο ομοιότητας

Δύο ταινίες ορίζονται όμοιες στο 9ο επίπεδο εάν έχουν:

 - τρια κοινά χαρακτηριστικά, μια κοινή αντιπροσωπευτική λέξη για την πλοκή τους, δύο κοινούς ηθοποιούς, έχουν κυκλοφορήσει την ίδια χρονική περίοδο ,είναι είτε έγχρωμες είτε ασπρόμαυρες, έχουν περίπου την ίδια χρονική διάρκεια, περίπου το ίδιο budget, την ίδια γλώσσα και έχουν διαδραματιστεί στην ίδια χώρα, ή

 - τρια κοινά χαρακτηριστικά, μια κοινή αντιπροσωπευτική λέξη για την πλοκή τους, κοινό σκηνοθέτη, έχουν κυκλοφορήσει την ίδια χρονική περίοδο ,είναι είτε έγχρωμες είτε ασπρόμαυρες, έχουν περίπου την ίδια χρονική διάρκεια, περίπου το ίδιο budget, την ίδια γλώσσα και έχουν διαδραματιστεί στην ίδια χώρα.

In [24]:
prolog.assertz('''(find_sim_9(X, Y):-
                                        identifier(Xid, X),
                                        same_genres_3(Xid, Yid), 
                                        same_plot_1(Xid, Yid),
                                        (same_actors_2(Xid, Yid) ; same_director(Xid, Yid)),
                                        same_year(Xid, Yid),
                                        same_color(Xid, Yid),
                                        same_duration(Xid, Yid),
                                        same_budget(Xid, Yid),
                                        same_country(Xid, Yid),
                                        same_language(Xid, Yid),
                                        identifier(Yid, Y))''')


### 10$^ο$ Επίπεδο ομοιότητας

Δύο ταινίες ορίζονται όμοιες στο 10ο επίπεδο εάν έχουν:

 - τρια κοινά χαρακτηριστικά, δύο κοινές αντιπροσωπευτικές λέξεις για την πλοκή τους, δύο κοινούς ηθοποιούς, μια κοινή εταιρία παραγωγής, κοινό σκηνοθέτη, έχουν κυκλοφορήσει την ίδια χρονική περίοδο, είναι είτε έγχρωμες είτε ασπρόμαυρες, έχουν περίπου την ίδια χρονική διάρκεια, έχουν περίπου το ίδιο budget, την ίδια γλώσσα, και έχουν διαδραματιστεί στην ίδια χώρα, ή

 - δύο κοινά χαρακτηριστικά, τρεις κοινές αντιπροσωπευτικές λέξεις για την πλοκή τους, δύο κοινούς ηθοποιούς, μια κοινή εταιρία παραγωγής, κοινό σκηνοθέτη, έχουν κυκλοφορήσει την ίδια χρονική περίοδο, είναι είτε έγχρωμες είτε ασπρόμαυρες, έχουν περίπου την ίδια χρονική διάρκεια, έχουν περίπου το ίδιο budget, την ίδια γλώσσα, και έχουν διαδραματιστεί στην ίδια χώρα.

In [25]:
prolog.assertz('''(find_sim_10(X, Y):-
                                        identifier(Xid, X),
                                        (same_genres_3(Xid, Yid), same_plot_2(Xid, Yid) ; same_plot_2(Xid, Yid), same_genres_3(Xid, Yid)),
                                        same_actors_2(Xid, Yid),
                                        same_production_company_1(Xid, Yid),
                                        same_director(Xid, Yid),
                                        same_year(Xid, Yid),
                                        same_color(Xid, Yid),
                                        same_duration(Xid, Yid),
                                        same_budget(Xid, Yid),
                                        same_country(Xid, Yid),
                                        same_language(Xid, Yid),
                                        identifier(Yid, Y))''')


### Ορισμός recommender

Έπειτα, ορίζουμε τη συνάρτηση <i>simple_recommender</i> η οποία δέχεται ως όρισμα μια ταινία και επιστρέφει ένα dictionary με keys τις τιμές από 0 έως 9 και values σύνολα με ταινίες όμοιες με την ταινία (που έχει περαστεί ως όρισμα στη συνάρτηση) σε επίπεδο από 1 έως 10 αντίστοιχα (κάθε ταινία εντάσσεται μια φορά στο dictionary στο μεγαλύτερο επίπεδο που μπορεί).

In [26]:
def recommender(movie):
    
    answers = {i : set() for i in range(0, 10)}
    s = set()
    
    
    q = prolog.query("find_sim_10('" + clean_text(movie) +"', M)")
    for soln in q:
        m = soln['M'] 
        if m not in s:
            s.add(soln['M'])
            answers[9].add(soln['M'])
    q.close()    
    
    
    q = prolog.query("find_sim_9('" + clean_text(movie) +"', M)")
    for soln in q:
        m = soln['M'] 
        if m not in s:
            s.add(soln['M'])
            answers[8].add(soln['M'])
    q.close()
    
    
    q = prolog.query("find_sim_8('" + clean_text(movie) +"', M)")
    for soln in q:
        m = soln['M'] 
        if m not in s:
            s.add(soln['M'])
            answers[7].add(soln['M'])
    q.close()    
    
    
    q = prolog.query("find_sim_7('" + clean_text(movie) +"', M)")
    for soln in q:
        m = soln['M'] 
        if m not in s:
            s.add(soln['M'])
            answers[6].add(soln['M'])
    q.close()
    
    
    q = prolog.query("find_sim_6('" + clean_text(movie) +"', M)")
    for soln in q:
        m = soln['M'] 
        if m not in s:
            s.add(soln['M'])
            answers[5].add(soln['M'])
    q.close()    
    
    
    q = prolog.query("find_sim_5('" + clean_text(movie) +"', M)")
    for soln in q:
        m = soln['M'] 
        if m not in s:
            s.add(soln['M'])
            answers[4].add(soln['M'])
    q.close()

    
    q = prolog.query("find_sim_4('" + clean_text(movie) +"', M)")
    for soln in q:
        m = soln['M'] 
        if m not in s:
            s.add(soln['M'])
            answers[3].add(soln['M'])
    q.close()    
    
    
    q = prolog.query("find_sim_3('" + clean_text(movie) +"', M)")
    for soln in q:
        m = soln['M'] 
        if m not in s:
            s.add(soln['M'])
            answers[2].add(soln['M'])
    q.close()    

    q = prolog.query("find_sim_2('" + clean_text(movie) +"', M)")
    for soln in q:
        m = soln['M'] 
        if m not in s:
            s.add(soln['M'])
            answers[1].add(soln['M'])
    q.close()    
    
    q = prolog.query("find_sim_1('" + clean_text(movie) +"',M)")
    for soln in q:
        m = soln['M'] 
        if m not in s:
            s.add(soln['M'])
            answers[0].add(soln['M'])
    q.close()    


    
    return answers

Παρακάτω, καλούμε ενδεικτικά το recommender που μόλις δημιουργήσαμε για τις ταινίες 'Avatar', 'Pirates of the Caribbean: At Worlds End' και 'The Avengers'.

In [27]:
recommender('Avatar')

{0: {'The Amazing Spider-Man 2',
  'Never Say Never Again',
  'This Means War',
  'Doctor Dolittle',
  'Hereafter',
  'Alien: Resurrection',
  'Lucy',
  'American Heist',
  'Everest',
  'Cirque du Soleil: Worlds Away',
  'Abduction',
  'Moonraker',
  'Bill & Teds Excellent Adventure',
  'Bad Boys',
  'One for the Money',
  'The Prince of Egypt',
  'Soldier',
  'Payback',
  'Flash Gordon',
  'Commando',
  'Being John Malkovich',
  'Ant-Man',
  'Taken 2',
  'Stone Cold',
  'Shipwrecked',
  'The Other Guys',
  'The Muppet Movie',
  'Lone Survivor',
  'Project Almanac',
  'Galaxina',
  'Spaced Invaders',
  'Money Talks',
  'Mission: Impossible - Rogue Nation',
  'Sheena',
  'Captain America: Civil War',
  'Daylight',
  'Young Sherlock Holmes',
  'Fight to the Finish',
  'Survival of the Dead',
  'Dawn of the Crescent Moon',
  'Beverly Hills Cop III',
  'G.I. Joe: Retaliation',
  'Lethal Weapon 3',
  'Days of Thunder',
  'El Mariachi',
  'Super',
  'Bait',
  'Falcon Rising',
  'The Peacemak

In [28]:
recommender('Pirates of the Caribbean: At Worlds End')

{0: {'127 Hours',
  '13 Hours: The Secret Soldiers of Benghazi',
  '1911',
  '1941',
  '2 Guns',
  '20,000 Leagues Under the Sea',
  '2001: A Space Odyssey',
  '21 Jump Street',
  '22 Jump Street',
  '3 Days to Kill',
  '3 Ninjas Kick Back',
  '30 Minutes or Less',
  '300: Rise of an Empire',
  '47 Ronin',
  '55 Days at Peking',
  'A Bugs Life',
  'A Good Day to Die Hard',
  'A Lonely Place to Die',
  'A Low Down Dirty Shame',
  'A Monster in Paris',
  'A Passage to India',
  'A Simple Wish',
  'A View to a Kill',
  'Abduction',
  'Abraham Lincoln: Vampire Hunter',
  'Ace Ventura: When Nature Calls',
  'Act of Valor',
  'Action Jackson',
  'After Earth',
  'Against the Wild',
  'Air Force One',
  'Airborne',
  'Airlift',
  'Aladdin',
  'Alex Cross',
  'Alice Through the Looking Glass',
  'Alice in Wonderland',
  'Alien',
  'Alien Zone',
  'Alien: Resurrection',
  'Aliens',
  'Alien³',
  'Alive',
  'All Is Lost',
  'Allegiant',
  'Alpha and Omega: The Legend of the Saw Tooth Cave',
  'A

In [29]:
recommender('The Avengers')

{0: {'Never Say Never Again',
  'Dungeons & Dragons: Wrath of the Dragon God',
  'Alien: Resurrection',
  'Ballistic: Ecks vs. Sever',
  'Dragon Wars: D-War',
  'Moonraker',
  'Bill & Teds Excellent Adventure',
  'Bad Boys',
  'Hulk',
  'Ice Age',
  'Soldier',
  'Payback',
  'The Prince of Egypt',
  'Half Past Dead',
  'Commando',
  'Hero',
  'Kiss Kiss Bang Bang',
  'Timecrimes',
  'Titan A.E.',
  'Flash Gordon',
  'Splice',
  'Stone Cold',
  'Shipwrecked',
  'The Muppet Movie',
  'Basic',
  'Galaxina',
  'Spider-Man',
  'The Dark Hours',
  'Kung Pow: Enter the Fist',
  'Repo! The Genetic Opera',
  'United 93',
  'Spaced Invaders',
  'Money Talks',
  'Sheena',
  'Dragonball Evolution',
  'Armored',
  'Daylight',
  'Sherlock Holmes',
  'Young Sherlock Holmes',
  'Into the Blue',
  'Harry Brown',
  'Hot Fuzz',
  'Beverly Hills Cop III',
  'X-Men: The Last Stand',
  'Whats the Worst That Could Happen?',
  'The Mexican',
  'Lethal Weapon 3',
  'Days of Thunder',
  'The Day After Tomorrow'

### Παρατήρηση:

Παρατηρούμε ότι στα μικρά επίπεδα ομοιότητας εμφανίζονται αρκετές ταινίες, κάτι το οποίο είναι επιθυμητό προκειμένου να στη συνέχεια (3ο μέρος) που θα φτιάξουμε το recommendation system να αποκτήσουμε "άποψη" έστω και μικρή όσες περισσότερες ταινίες μπορούμε για έναν συγκεκριμένο χρήστη, έτσι ώστε να μπορούμε να κάνουμε την καλύτερη δυνατή και πιο πλήρη επιλογή για αυτόν.  

# **3ο Μέρος: Recommendation System Με βάση τις προτιμήσεις-Αξιολογήσεις του χρήστη-Εκπαίδευση και Πρόβλεψη**


Αρχικά μελετάμε τις αξιολογήσεις κάθε χρήστη για να καταλάβουμε την δομή και τις πληροφορίες κάθε αρχείου.




Μέσω της συνάρτησης train_recommender εκπαιδεύουμε τον recommender με βάση το train set, δηλαδή τις αξιολογήσεις ενός συγκεκριμένου χρήστη. Πιο αναλυτικά, για όλες τις ταινίες που αυτος έχει αξιολογήσει (με βαθμολογία από 0 έως 5) αντιστοιχίζουμε στο recommender ένα βάρος σε κάθε όμοια ταινία με αυτή, με βάση τη βαθμολογία που έβαλε ο συγκεκριμένος χρήστης και το επίπεδο ομοιότητας στο οποίο οι δύο ταινίες εντάσσονται. Έτσι, για ταινίες που ο χρήστης έχει βαθμολογήσει με 0 ή 1 επιλέγουμε αρνητικά βάρη, καθώς θεωρούμε ότι ο χρήστης προτιμάει και προτίνουμε σε αυτόν ταινίες όμοιες με ταινία που έχει βαθμολογήσει με 2 και πάνω, ενώ παράλληλα όσο πιο όμοια είναι μια ταινία με αυτή τόσο περισσότερο αυξάνουμε το θετικό (βαθμολογία από 2 έως 5) ή αρνητικό της βάρος (βαθμολογία από 0 έως 1). Με τον τρόπο αυτό μετά την εκπαίδευση του recommender θα προτίνουμε στο χρήστη ταινίες που έχουν συγκεντρώσει θετικό score (βάρος).  
Έπειτα, μέσω της συνάρτησης predict προβλέπουμε, με βάση τα score για κάθε ταινία που υπολογίσαμε παραπάνω, ποιές από τις ταινίες αυτές θα προτιμούσε ο χρήστης, οπότε και τις προτίνουμε και τελικά χρησιμοποιούμε τη συνάρτηση get_metrics για να δούμε με βάση το test set πόσες από τις ταινίες που πρότινε ο recommender συγκεκριμένο χρήστη πράγματι προτιμούνται από αυτόν. Αυτό το επιτυγχάνουμε μέσω των μετρικών precision, recall και F1. 

In [30]:
from tqdm.notebook import tqdm
from sklearn.metrics import precision_score, recall_score, f1_score


rating_weights = {0: -1, 1: -0.5, 2:0, 3:0.3, 4:0.7, 5:1}
score_weights = {i:i + 1 for i in range(0, 8)} # ανάλογα με τα επίπεδα ομοιότητας που έχουν οριστεί στην simple_recommender

def train_recommender(ratings, rating_weights, score_weights, start=0, end = -1, movie_score = {}, simple = 1):
    """
    Στην συνάρτηση αυτή μπορούμε να ορίζουμε ποιο υποσύνολο των αξιολογήσεων θα χρησιμοποιήσουμε για το train,
    ενώ παράλληλα μπορούμε να αρχικοποιήσουμε και το movie_score το οποίο αποθηκεύονται τα score κάθε ταινίας.
    Έτσι αν έχουμε κάνει train τον recommender μας για τις πρώτες 10 ταινίες, μπορούμε να συνεχίσουμε για τις 
    υπόλοιπες χωρίς κάθε φορά να ξανά-διαμορφώνουμε το score για αυτές. Αυτό θα σας βοηθήσει στην εκτέλεση και χρονικά όταν ελέγχετε την 
    απόδοση σε υποσύνολα των ratings π.χ. πρώτες 10 ταινίες έπειτα τις επόμενες 20 κ.ο.κ.
    """
    if end == -1:
        end = len(ratings)

    l = 8
    
    if simple == 0:
        score_weights[8] = 9
        score_weights[9] = 10
        l = 10
        
    ratings = ratings.iloc[range(start, end)]
    for row in tqdm(ratings.itertuples(index=True, name='Pandas')):
        movie = clean_text(getattr(row, 'movie_title'))
        rating = getattr(row, 'rating')

        if simple == 1:
            similar_movies = simple_recommender(movie)
        else :
            similar_movies = recommender(movie)
            
        for i in range(0, l):
            for similar_movie in similar_movies[i]:
                if similar_movie not in movie_score:
                    movie_score[similar_movie] = rating_weights[int(rating)] * score_weights[i]
                else:
                    movie_score[similar_movie] += rating_weights[int(rating)] * score_weights[i] #το weight θα το ορίσετε ανα επίπεδο ομοιότητας οι πολύ όμοιες ταινίες θα έχουν μεγαλύτερο βάρος
    return movie_score


#αυτό είναι ένα παράδειγμα για το πως θα μπορούσε να υλοποιήθει η predict
#έχουμε ορίσει ότι μια ταινία θα έπρεπε να είναι προτεινόμενη αν είχε σκορ > 0
def predict(ratings, movie_score):
    real, pred = [], []
    for i, row in enumerate(ratings.itertuples(index=True, name='Pandas')):
        movie = clean_text(getattr(row, 'movie_title'))
        rating = getattr(row, 'rating')

        if movie in movie_score: #αν έχουμε σχηματίσει βαθμολογία για την ταινία αυτή
            pred.append(int(movie_score[movie] > 0)) #heuristic για το αν μια ταινία είναι προτεινόμενη
            real.append(int(rating > 3))# έτσι ορίζουμε ότι μια ταινία θα έπρεπε να είναι προτεινόμενη
            #η συνθήκη αυτή δεν μπορεί να αλλάξει
        else: #δεν μπορούμε να προτείνουμε κάτι για το όποιο δεν έχουμε σχηματίσει εικόνα
            pred.append(0)
            real.append(int(rating > 3))

    return real, pred


def get_metrics(real, pred):
    metrics = {}
    metrics["precision"] = precision_score(real, pred)
    metrics["recall"] = recall_score(real, pred)
    metrics["f1"] = f1_score(real, pred)
    return metrics

In [28]:
train_ratings = pd.read_csv("train_ratings.csv")
test_ratings = pd.read_csv("test_ratings.csv")

In [30]:
movie_score = train_recommender(train_ratings, rating_weights, score_weights, 0, 10, simple = 0)
real, pred = predict(test_ratings, movie_score)
print (get_metrics(real, pred))


{'precision': 0.4957983193277311, 'recall': 0.8194444444444444, 'f1': 0.6178010471204188}


In [32]:
movie_score = train_recommender(train_ratings, rating_weights, score_weights, 10, 20, movie_score, simple = 0)
real, pred = predict(test_ratings, movie_score)
print (get_metrics(real, pred))


{'precision': 0.5073529411764706, 'recall': 0.9583333333333334, 'f1': 0.6634615384615384}


In [33]:
movie_score = train_recommender(train_ratings, rating_weights, score_weights, 20, 30, movie_score, simple = 0)
real, pred = predict(test_ratings, movie_score)
print (get_metrics(real, pred))


{'precision': 0.5071428571428571, 'recall': 0.9861111111111112, 'f1': 0.6698113207547169}


In [34]:
movie_score = train_recommender(train_ratings, rating_weights, score_weights, 30, 50, movie_score, simple = 0)
real, pred = predict(test_ratings, movie_score)
print (get_metrics(real, pred))


{'precision': 0.5071428571428571, 'recall': 0.9861111111111112, 'f1': 0.6698113207547169}


In [ ]:
movie_score = train_recommender(train_ratings, rating_weights, score_weights,start = 50, movie_score= movie_score)
real, pred = predict(test_ratings, movie_score)
print (get_metrics(real, pred))

### Παρατηρήσεις:

<b> 1. </b> Μετά από δοκιμές στο recommendation system επιλέξαμε ως βάρη γα τις αξιολογήσεις τα ακόλουθα:

    - rate 0: -1
    - rate 1: -0.5
    - rate 2: 0
    - rate 3: 0.3
    - rate 4: 0.7
    - rate 5: 1
καθώς παρατηρήσαμε ότι αυτά μας εξασφαλίζουν όσο το δυνατόν γρηγορότερα ένα καλύτερο recommndation system με βάση τις μετρικές. Η επιλογή αυτή είναι αρκετά συμβατή με την πραγματικότητα εάν αναλογιστεί κανείς ότι οι περισσότεροι που καλούνται να αξιολογήσουν μια ταινία με κλίμακα 0 έως 5, συνήθως επιλέγουν το 0 ή το 1 εάν η ταινία σχεδόν τους απογοήτευσε, το 2 εάν είναι κάπως ουδέτεροι ενώ το 3, 4 ή 5 εάν του άρεσε (με την προφανή κλιμάκωση) οπότε και πιθανόν θα επέλεγαν να δουν μια παρόμοια ταινία. Βέβαια, το ότι στο συγκεκριμενο παράδειγμα η επιλογή των rating αυτών δουλεύει αρκετά καλά ίσως είναι και λίγο υποκειμενικό καθώς εργαζόμαστε πάνω σε ένα συγκεκριμένο χρήστη, οπότε εδώ παρατηρούμε ότι τα ratings αυτά συμβαδίζουν αρκετά με τις προτιμήσεις του συγκεκριμένου χρήστη ενώ για παράδειγμα κάποιος άλλος χρήστης μπορεί να επέλεγε την αξιολόγηση 3 για κάποια ταινία που του πέρασε σχετικά αδιάφορή (οπότε και εκεί ίσως χρειαζόταν μικρότερο βάρος για την αξιολόγηση αυτή).

<b> 2. </b> Παρατηρούμε ότι από τις πρώτες δέκα ακόμα αξιολογήσεις, το recommndation σύστημα που κατασκευάσαμε εκπαιδεύεται σε αρκετά καλό βαθμό, τουλάχιστον για το συγκεκριμένο χρήστη, πετυχαίνοντας να προβλέψει σωστά τις ταινίες που θα προτιμούσε πράγματι να δει σε ποσοστό 81% (recall). Στη συνέχεια μάλιστα παρατηρούμε ότι μόλις με δέκα ακόμα αξιολογήσεις το ποσοστό των ταινιών που προβλέπει σωστά το σύστημα αυξάνεται σε 95% φτάνοντας αρκετά γρήγορα (με ένα μικρό train set) σε μια αρκετά ικακνοποιητική τιμή. Έπειτα, όσο συνεχίζεται η εκπαίδευση του συστήματος, η μετρική recall αυξάνεται όλο καιιι περσσότερο, ενώ παρατηρούμε ότι μετά τις 30 πρώτες αξιολογήσεις το σύστημα ήδη έχει εκπαιδευτεί σε αρκετά καλό βαθμό, οπότε και η αύξηση αυτής γίνεται πλέον αρκετά αργά.   

<b> 3.</b> Όσον αφορά τη μετρική precission παρατηρούμε ότι και αυτή καθώς εκπαιδεύεται το σύστημα αυξάνεται, ωστόσο με αρκετά αργό ρυθμό τουλάχιστον για το μικρό αυτό πλήθος αξιολογήσεων με το οποίο σταδιακά εκπαιδεύουμε το σύστημα μας.

<b> 4. </b> Παρατηρούμε ότι η εκπαίδευση του συστήματος με τις τελευταίες 50 αξιολογήσεις αποτυγχάνει. Οπότε, <b>θα προσπαθήσουμε να μετριάσουμε λίγο την πολυπλοκότητα του συστήματος μας, μειώνοντας τα επίπεδα διάκρισης ομοιότητας μεταξύ των ταινιών έτσι ώστε να περιορίσουμε λίγο τη μνήμη και τον υπολογιστικό χρόνο που απαιτείται για την εκπαίδευη του recommendation και στη συνέχεια θα δούμε εάν η τροποποίηση αυτή έχει επίπτωση στην απόδοση του συστήματος</b>.  

## Ένα πιο απλό recommendation system

### 1$^ο$ Επίπεδο ομοιότητας

Δύο ταινίες ορίζονται όμοιες στο 1ο επίπεδο εάν έχουν ένα κοινό χαρακτηριστικό.

In [35]:
prolog.assertz('''(simple_find_sim_1(X, Y):-
                                        identifier(Xid, X),
                                        same_genres_1(Xid, Yid),
                                        identifier(Yid, Y))''')

### 2$^ο$ Επίπεδο ομοιότητας

Δύο ταινίες ορίζονται όμοιες στο 2ο επίπεδο εάν έχουν ένα κοινό χαρακτηριστικό, μια κοινή αντιπροσωπευτική λέξη για την πλοκή τους, έχουν κυκλοφορήσει την ίδια χρονική περίοδο και είναι είτε έγχρωμες είτε ασπρόμαυρες.

In [36]:
prolog.assertz('''(simple_find_sim_2(X, Y):-
                                        identifier(Xid, X),
                                        same_genres_1(Xid, Yid),
                                        same_plot_1(Xid, Yid),
                                        same_year(Xid, Yid),
                                        same_color(Xid, Yid),
                                        identifier(Yid, Y))''')

### 3$^ο$ Επίπεδο ομοιότητας

Δύο ταινίες ορίζονται όμοιες στο 3ο επίπεδο εάν έχουν ένα κοινό χαρακτηριστικό, μια κοινή αντιπροσωπευτική λέξη για την πλοκή τους, έναν κοινό ηθοποιό, έχουν κυκλοφορήσει την ίδια χρονική περίοδο και είναι είτε έγχρωμες είτε ασπρόμαυρες.

In [37]:
prolog.assertz('''(simple_find_sim_3(X, Y):-
                                        identifier(Xid, X),
                                        same_genres_1(Xid, Yid),
                                        same_plot_1(Xid, Yid),
                                        same_actors_1(Xid, Yid),
                                        same_year(Xid, Yid),
                                        same_color(Xid, Yid),
                                        identifier(Yid, Y))''')

### 4$^ο$ Επίπεδο ομοιότητας

Δύο ταινίες ορίζονται όμοιες στο 4ο επίπεδο εάν έχουν:

- δύο κοινά χαρακτηριστικά, μια κοινή αντιπροσωπευτική λέξη για την πλοκή τους, έχουν κυκλοφορήσει την ίδια χρονική περίοδο, είναι είτε έγχρωμες είτε ασπρόμαυρες και έχουν περίπου την ίδια χρονική διάρκεια, ή

- δύο κοινές αντιπροσωπευτικές λέξεις για την πλοκή τους, ένα κοινό χρακτηριστικό, έχουν κυκλοφορήσει την ίδια χρονική περίοδο, είναι είτε έγχρωμες είτε ασπρόμαυρες και έχουν περίπου την ίδια χρονική διάρκεια.

In [38]:
prolog.assertz('''(simple_find_sim_4(X, Y):-
                                        identifier(Xid, X),
                                        (same_genres_2(Xid, Yid), same_plot_1(Xid, Yid) ; same_genres_1(Xid, Yid), same_plot_2(Xid, Yid)),
                                        same_year(Xid, Yid),
                                        same_duration(Xid, Yid),
                                        same_color(Xid, Yid),
                                        identifier(Yid, Y))''')

### 5$^ο$ Επίπεδο ομοιότητας

Δύο ταινίες ορίζονται όμοιες στο 5ο επίπεδο εάν έχουν:

- δύο κοινά χαρακτηριστικά, μια κοινή αντιπροσωπευτική λέξη για την πλοκή τους, έναν κοινό ηθοποιό, έχουν κυκλοφορήσει την ίδια χρονική περίοδο, είναι είτε έγχρωμες είτε ασπρόμαυρες και έχουν περίπου την ίδια χρονική διάρκεια, ή

- δύο κοινά χαρακτηριστικά, μια κοινή αντιπροσωπευτική λέξη για την πλοκή τους, μια κοινή εταιρία παραγωγής, έχουν κυκλοφορήσει την ίδια χρονική περίοδο, είναι είτε έγχρωμες είτε ασπρόμαυρες και έχουν περίπου την ίδια χρονική διάρκεια.

In [39]:
prolog.assertz('''(simple_find_sim_5(X, Y):-
                                        identifier(Xid, X),
                                        same_genres_2(Xid, Yid), 
                                        same_plot_1(Xid, Yid),
                                        (same_actors_1(Xid, Yid) ; same_production_company_1(Xid, Yid)),
                                        same_year(Xid, Yid),
                                        same_duration(Xid, Yid),
                                        same_color(Xid, Yid),
                                        identifier(Yid, Y))''')

### 6$^ο$ Επίπεδο ομοιότητας

Δύο ταινίες ορίζονται όμοιες στο 5ο επίπεδο εάν έχουν:

- τρια κοινά χαρακτηριστικά, έναν κοινό ηθοποιό, έχουν κυκλοφορήσει την ίδια χρονική περίοδο, είναι είτε έγχρωμες είτε ασπρόμαυρες και έχουν περίπου την ίδια χρονική διάρκεια, ή

- δύο κοινά χαρακτηριστικά, δύο κοινές αντιπροσωπευτικές λέξεις για την πλοκή τους, μια κοινή εταιρία παραγωγής, έχουν κυκλοφορήσει την ίδια χρονική περίοδο, είναι είτε έγχρωμες είτε ασπρόμαυρες και έχουν περίπου την ίδια χρονική διάρκεια.

In [40]:
prolog.assertz('''(simple_find_sim_6(X, Y):-
                                        identifier(Xid, X),
                                        (same_genres_3(Xid, Yid) ; same_genres_2(Xid, Yid), same_plot_2(Xid, Yid)),
                                        same_actors_1(Xid, Yid),
                                        same_year(Xid, Yid),
                                        same_duration(Xid, Yid),
                                        same_color(Xid, Yid),
                                        identifier(Yid, Y))''')

### 7$^ο$ Επίπεδο ομοιότητας

Δύο ταινίες ορίζονται όμοιες στο 5ο επίπεδο εάν έχουν:

- τρια κοινά χαρακτηριστικά, κοινό σκηνοθέτη, έχουν κυκλοφορήσει την ίδια χρονική περίοδο, είναι είτε έγχρωμες είτε ασπρόμαυρες και έχουν περίπου την ίδια χρονική διάρκεια, ή

- δύο κοινά χαρακτηριστικά, δύο κοινές αντιπροσωπευτικές λέξεις για την πλοκή τους, κοινό σκηνοθέτη, έχουν κυκλοφορήσει την ίδια χρονική περίοδο, είναι είτε έγχρωμες είτε ασπρόμαυρες και έχουν περίπου την ίδια χρονική διάρκεια.

In [41]:
prolog.assertz('''(simple_find_sim_7(X, Y):-
                                        identifier(Xid, X),
                                        (same_genres_3(Xid, Yid) ; same_genres_2(Xid, Yid), same_plot_2(Xid, Yid)),
                                        same_director(Xid, Yid),
                                        same_year(Xid, Yid),
                                        same_duration(Xid, Yid),
                                        same_color(Xid, Yid),
                                        identifier(Yid, Y))''')

### 8$^ο$ Επίπεδο ομοιότητας

Δύο ταινίες ορίζονται όμοιες στο 5ο επίπεδο εάν έχουν τρια κοινά χαρακτηριστικά, δύο κοινούς ηθοποιούς, κοινό σκηνοθέτη, μια κοινή εταιρία παραγωγής, έχουν κυκλοφορήσει την ίδια χρονική περίοδο, είναι είτε έγχρωμες είτε ασπρόμαυρες και έχουν περίπου την ίδια χρονική διάρκεια.

In [42]:
prolog.assertz('''(simple_find_sim_8(X, Y):-
                                        identifier(Xid, X),
                                        same_genres_3(Xid, Yid),
                                        same_actors_2(Xid, Yid),
                                        same_production_company_1(Xid, Yid),
                                        same_director(Xid, Yid),
                                        same_year(Xid, Yid),
                                        same_duration(Xid, Yid),
                                        same_color(Xid, Yid),
                                        identifier(Yid, Y))''')

In [43]:
def simple_recommender(movie):
    
    answers = {i : set() for i in range(0, 8)}
    s = set()
    
    
    q = prolog.query("simple_find_sim_8('" + clean_text(movie) +"', M)")
    for soln in q:
        m = soln['M'] 
        if m not in s:
            s.add(soln['M'])
            answers[7].add(soln['M'])
    q.close()    
    
    
    q = prolog.query("simple_find_sim_7('" + clean_text(movie) +"', M)")
    for soln in q:
        m = soln['M'] 
        if m not in s:
            s.add(soln['M'])
            answers[6].add(soln['M'])
    q.close()
    
    
    q = prolog.query("simple_find_sim_6('" + clean_text(movie) +"', M)")
    for soln in q:
        m = soln['M'] 
        if m not in s:
            s.add(soln['M'])
            answers[5].add(soln['M'])
    q.close()    
    
    
    q = prolog.query("simple_find_sim_5('" + clean_text(movie) +"', M)")
    for soln in q:
        m = soln['M'] 
        if m not in s:
            s.add(soln['M'])
            answers[4].add(soln['M'])
    q.close()

    
    q = prolog.query("simple_find_sim_4('" + clean_text(movie) +"', M)")
    for soln in q:
        m = soln['M'] 
        if m not in s:
            s.add(soln['M'])
            answers[3].add(soln['M'])
    q.close()    
    
    
    q = prolog.query("simple_find_sim_3('" + clean_text(movie) +"', M)")
    for soln in q:
        m = soln['M'] 
        if m not in s:
            s.add(soln['M'])
            answers[2].add(soln['M'])
    q.close()    

    q = prolog.query("simple_find_sim_2('" + clean_text(movie) +"', M)")
    for soln in q:
        m = soln['M'] 
        if m not in s:
            s.add(soln['M'])
            answers[1].add(soln['M'])
    q.close()    
    
    q = prolog.query("simple_find_sim_1('" + clean_text(movie) +"',M)")
    for soln in q:
        m = soln['M'] 
        if m not in s:
            s.add(soln['M'])
            answers[0].add(soln['M'])
    q.close()    


    
    return answers

Καλούμε ενδεικτικά το recommender που μόλις δημιουργήσαμε για τις ταινίες 'Avatar', 'Pirates of the Caribbean: At Worlds End' και 'The Avengers'.

In [44]:
simple_recommender('Avatar')

{0: {'Medicine Man',
  'Timecrimes',
  'Asterix at the Olympic Games',
  'Ip Man 3',
  'Austin Powers in Goldmember',
  'Ella Enchanted',
  'The Masked Saint',
  'Life of Pi',
  'Killing Zoe',
  'Touching the Void',
  'You Only Live Twice',
  'Enders Game',
  'Carriers',
  'Bloody Sunday',
  'Mr. Peabody & Sherman',
  'The World Is Not Enough',
  'Aladdin',
  'Das Boot',
  'The Warriors Way',
  'Days of Thunder',
  'G.I. Jane',
  'Mighty Joe Young',
  'The BFG',
  'Gremlins',
  'Blade Runner',
  'Epic',
  'Little Nicky',
  'Con Air',
  'Captain Phillips',
  'I Origins',
  'Boom Town',
  'The Last Days on Mars',
  'Smokin Aces',
  'Showdown in Little Tokyo',
  'Charlies Angels',
  'The Wizard of Oz',
  'Six-String Samurai',
  'Green Lantern',
  'Duplex',
  'Re-Kill',
  'The Flintstones',
  'The Mask',
  'Pacific Rim',
  'The Lawnmower Man',
  'Micmacs',
  'Nomad: The Warrior',
  'Sleeper',
  'Hocus Pocus',
  'The Long Kiss Goodnight',
  'Quest for Camelot',
  'The Scorpion King',
  'Alp

In [45]:
recommender('Pirates of the Caribbean: At Worlds End')

{0: {'127 Hours',
  '13 Hours: The Secret Soldiers of Benghazi',
  '1911',
  '1941',
  '2 Guns',
  '20,000 Leagues Under the Sea',
  '2001: A Space Odyssey',
  '21 Jump Street',
  '22 Jump Street',
  '3 Days to Kill',
  '3 Ninjas Kick Back',
  '30 Minutes or Less',
  '300: Rise of an Empire',
  '47 Ronin',
  '55 Days at Peking',
  'A Bugs Life',
  'A Good Day to Die Hard',
  'A Lonely Place to Die',
  'A Low Down Dirty Shame',
  'A Monster in Paris',
  'A Passage to India',
  'A Simple Wish',
  'A View to a Kill',
  'Abduction',
  'Abraham Lincoln: Vampire Hunter',
  'Ace Ventura: When Nature Calls',
  'Act of Valor',
  'Action Jackson',
  'After Earth',
  'Against the Wild',
  'Air Force One',
  'Airborne',
  'Airlift',
  'Aladdin',
  'Alex Cross',
  'Alice Through the Looking Glass',
  'Alice in Wonderland',
  'Alien',
  'Alien Zone',
  'Alien: Resurrection',
  'Aliens',
  'Alien³',
  'Alive',
  'All Is Lost',
  'Allegiant',
  'Alpha and Omega: The Legend of the Saw Tooth Cave',
  'A

In [46]:
recommender('The Avengers')

{0: {'Bee Movie',
  'Escape from New York',
  'The Pink Panther',
  'The Incredibles',
  'Medicine Man',
  'Timecrimes',
  'The Matrix Revolutions',
  'Asterix at the Olympic Games',
  'The Emperors New Groove',
  'Jonah: A VeggieTales Movie',
  'Somewhere in Time',
  'Final Fantasy: The Spirits Within',
  'Hellboy',
  'Reno 911!: Miami',
  'Thomas and the Magic Railroad',
  'Laws of Attraction',
  'Superman',
  'The Last of the Mohicans',
  'Big Trouble',
  'The Black Stallion',
  'Austin Powers in Goldmember',
  'Memoirs of an Invisible Man',
  'Men in Black',
  'The Rundown',
  'Gun Shy',
  'Killing Zoe',
  'Nutty Professor II: The Klumps',
  'A.I. Artificial Intelligence',
  'Spy Kids 2: The Island of Lost Dreams',
  'The Corruptor',
  'Executive Decision',
  'Touching the Void',
  'You Only Live Twice',
  'All Hat',
  'Race to Witch Mountain',
  'Carriers',
  'Hollywood Homicide',
  'Megiddo: The Omega Code 2',
  'The Messenger: The Story of Joan of Arc',
  'Bloody Sunday',
  'The

Αφού δημιουργήσαμε έναν πιο απλό recommender ταινιών θα προσπαθήσουμε δημιουργήσουμε και να εκπαιδεύσουμε με το ίδιο test set ένα recommendation system (βασισμένο στο simple recommender) και θα δούμε κατά πόσο αυτό επιτυγχάνει να εντοπίσει και να προτίνει σωστά τις ταινίες που προτιμά ο συγκεκριμένος χρήστης.

In [26]:
train_ratings = pd.read_csv("train_ratings.csv")
test_ratings = pd.read_csv("test_ratings.csv")

In [27]:
movie_score = train_recommender(train_ratings, rating_weights, score_weights, 0, 10)
real, pred = predict(test_ratings, movie_score)
print (get_metrics(real, pred))


{'precision': 0.4921875, 'recall': 0.875, 'f1': 0.63}


In [28]:
movie_score = train_recommender(train_ratings, rating_weights, score_weights, 10, 20, movie_score)
real, pred = predict(test_ratings, movie_score)
print (get_metrics(real, pred))


{'precision': 0.5074626865671642, 'recall': 0.9444444444444444, 'f1': 0.6601941747572816}


In [29]:
movie_score = train_recommender(train_ratings, rating_weights, score_weights, 30, 50, movie_score)
real, pred = predict(test_ratings, movie_score)
print (get_metrics(real, pred))


{'precision': 0.5072463768115942, 'recall': 0.9722222222222222, 'f1': 0.6666666666666666}


In [30]:
movie_score = train_recommender(train_ratings, rating_weights, score_weights,start = 50, movie_score= movie_score)
real, pred = predict(test_ratings, movie_score)
print (get_metrics(real, pred))


{'precision': 0.5106382978723404, 'recall': 1.0, 'f1': 0.676056338028169}


### Παρατηρήσεις:

<b> 1. </b> Αρχικά, παρατηρούμε ότι η εκπαίδευση του recommendation system πάνω σε όλες τις αξιολογήσεις του test set είναι επιτυχής, μετά από την απλοποίηση του recommender.

<b> 2. </b> Επιπλέον παρατηρούμε ότι συγκριτικά με το recommender δέκα επιπέδων ομοιότητας, ο πιο απλός recommender οκτώ επιπέδων που μόλις κατασκευάσαμε δεν καταφέρνει να εντοπίσει τόσο καλά με τις πρώτες αξιολογήσεις τις προτιμήσεις του συγκεκριμένου χρήστη. Ωστόσο με την εκπαίδευση πάνω σε περισσότερες παρατηρήσεις οι μετρικές του συστήματος βελτιώνονται γρήγορα και με τις πρώτες 50 αξιολογήσεις το recommendation system καταφέρνει να προβλέψει σωστά τις ταινίες που θα προτιμούσε πράγματι να δει ο χρήστης σε ποσοστό 97% (recall). Παράλληλα, η μετρική precision παρατηρούμε ότι και πάλι αυξάνεται ελάχιστα με αργό ρυθμό, όπως και στο recommendation sytem με δέκα επίπεδα ομοιότητας. 

<b>Οπότε, τελικά μπορούμε να πούμε ότι η τροποποίηση του recommender μπορεί να μείωσε ελάχιστα την ταχύτητα με την οποία το recommendation system, μέσω της εκπαίδευσης, φτάνει σε μια πολύ καλή επιθυμητή απόδοση απαιτώντας εκπαίδευση σε ένα λίγο μεγαλύτερο train set αλλά πρακτικά βοήθησε αρκετά στο να μειωθεί η πολυπλοκότητα, η απαιτούμενη μνήμη και ο υπολογιστικός χρόνος του συστήματος, οπότε και είναι προτιμητέα </b>.    